hphp/util/asm-x64-intelxed.h

```diff
@@ -0,0 +1,1031 @@
+  }
+
+  inline xed_reg_enum_t xedFromReg (const Reg8& reg) {
+    int regid = int(reg);
+    if ((regid & 0x80) == 0) {
+      return xed_reg_enum_t(regid + XED_REG_AL);
+    }
+    return xed_reg_enum_t((regid - 0x84) + XED_REG_AH);
+  }
+
+  inline xed_reg_enum_t xedFromReg (const RegXMM& reg) {
+    return xed_reg_enum_t(int(reg) + XED_REG_XMM0);
+  }
+
+  static int getDisplSize(intptr_t value) {
+    if (value == 0) {
+      return sz::nosize;
+    }
+    return deltaFits(value, sz::byte) ? sz::byte : sz::dword;
+  }
+
+  xed_enc_displacement_t xedDispFromValue(intptr_t value, int size) {
+    switch(size) {
+      case sz::nosize:  return {0, 0};
+      case sz::byte:    return {(xed_uint64_t)(safe_cast<int8_t>(value)),
+                                SZ_TO_BITS((xed_uint32_t) size)};
+      default:          return {(xed_uint64_t)(safe_cast<int32_t>(value)),
+                                SZ_TO_BITS((xed_uint32_t) size)};
+    }
+  }
+
+  xed_enc_displacement_t xedDispFromValue(intptr_t value) {
+    return xedDispFromValue(value, getDisplSize(value));
+  }
+
+  template<typename regtype>
+  explicit XedOperand(const regtype& reg) {
+    op = xed_reg(xedFromReg(reg));
+  }
+
+  explicit XedOperand(xed_reg_enum_t reg) {
+    op = xed_reg(reg);
+  }
+
+  explicit XedOperand(const MemoryRef& m, int memSize) {
+    xed_reg_enum_t base = (int(m.r.base) != -1 ?
+                            xedFromReg(m.r.base) : XED_REG_INVALID);
+    xed_reg_enum_t index = (int(m.r.index) != -1 ?
+                            xedFromReg(m.r.index) : XED_REG_INVALID);
+    op = xed_mem_gbisd(xed_reg_enum_t(m.segment), base, index, m.r.scale,
+                       xedDispFromValue(m.r.disp), SZ_TO_BITS(memSize));
+  }
+
+  explicit XedOperand(const RIPRelativeRef& r, int memSize) {
+    op = xed_mem_bd(XED_REG_RIP, xedDispFromValue(r.r.disp, sz::dword),
+                    SZ_TO_BITS(memSize));
+  }
+
+  template<typename immtype>
+  explicit XedOperand(const immtype& immed, int immSize) {
+    op = xed_imm0(xed_imm_value(immed, immSize).uq, SZ_TO_BITS(immSize));
+  }
+
+  explicit XedOperand(CodeAddress address, int size) {
+    int64_t target = (int64_t)address;
+    assert(deltaFits(target, size) &&
+           "Relative address doesn't fit selected size");
+    op = xed_relbr((int32_t)target, SZ_TO_BITS(size));
+  }
+
+  template<typename immtype>
+  explicit XedOperand(const immtype& immed, int immSizes, immFitFunc func) {
+    immSizes = reduceImmSize(immed.q(), immSizes, func);
+    op = xed_imm0(xed_imm_value(immed, immSizes).uq, SZ_TO_BITS(immSizes));
+  }
+
+  inline int reduceImmSize(int64_t value, int allowedSizes, immFitFunc func) {
+    for (int crtSize = sz::byte; crtSize < sz::qword; crtSize <<= 1) {
+      if ((allowedSizes & crtSize) && (*func)(value, crtSize))
+        return crtSize;
+      }
+    assert((allowedSizes & sz::qword) &&
+           "Could not find an optimal size for Immed");
+    return sz::qword;
+  }
```

Comment:
Something is off with the indentation and/or the block-ending `}` between 163 and here.